File: 215.kth-largest-element-in-an-array.cpp
/*
 * @lc app=leetcode id=215 lang=cpp
 *
 * [215] Kth Largest Element in an Array
 */

#include "leetcode.h"

// @lc code=start
class Solution {
public:
    /*
    partition like quick sort:
    1 2 3 4
    4 3 2 1
    1 4 6 5
    */
    int findKthLargest(vector<int>& nums, int k) {
        int l = 0, r = nums.size() - 1;        
        while (l < r) {
            // use last element to partition
            int s = l;
            int t = r;
            while (l < r) {
                // 不能取等号，防止较大的元素也放到左边了。
                while (l < r && nums[l] < nums[t]) l++;
                while (l < r && nums[r] >= nums[t]) r--;
                swap(nums[l], nums[r]);
                // 不能移动，防止结束时指向一个较小元素，然后后面swap导致这个较小元素被交换到了右边。
                // l++; r--;
            }
            // l == r
            swap(nums[l], nums[t]);

            if (t - l + 1 == k) return nums[l];
            if (t - l + 1 > k) {
                l++;
                r = t;
            } else {
                k -= t - l + 1;
                l = s;
                r--;
            }
        }
        // l == r
        return nums[l];
    }
};
// @lc code=end

File: go/215.kth-largest-element-in-an-array.go
/*
 * @lc app=leetcode id=215 lang=go
 *
 * [215] Kth Largest Element in an Array
 */

package main

// @lc code=start
func findKthLargest(nums []int, k int) int {
	l, r := 0, len(nums)-1
	for l < r {
		// use last element to partition
		s := l
		t := r
		for l < r {
			for l < r && nums[l] < nums[t] {
				l++
			}
			for l < r && nums[r] >= nums[t] {
				r--
			}
			nums[l], nums[r] = nums[r], nums[l]
		}
		// l == r
		nums[l], nums[t] = nums[t], nums[l]
		if t-l+1 == k {
			return nums[l]
		}
		if t-l+1 > k {
			l++
			r = t
		} else {
			k -= t - l + 1
			l = s
			r--
		}
	}
	// l == r
	return nums[l]
}

// @lc code=end

func main() {
	findKthLargest([]int{3, 2, 3, 1, 2, 4, 5, 5, 6, 7, 7, 8, 2, 3, 1, 1, 1, 10, 11, 5, 6, 2, 4, 7, 8, 5, 6}, 2)
}

/*
[3,2,3,1,2,4,5,5,6,7,7,8,2,3,1,1,1,10,11,5,6,2,4,7,8,5,6]
[3,2,3,1,2,4,5,5,5,4,2,5,2,3,1,1,1,6,  11,8,6,7,7,7,8,6,10] 6,8,6,7,7,7,8,11,10
[3,2,3,1,2,4,5,5,5,4,2,5,2,3,1,1,1,6,  6,8,6,7,7,7,10,11,8]
*/
